### PR TITLE
Introduce Postgres queue

### PR DESCRIPTION
**Pull request by vchan**
This PR introduces a Postgres-backed queue. Still need to add tests and implement several missing features like info, stats, schedule, sweep, etc.

### CODE REVIEW

File: saq/queue/postgres.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        LoadType,
+        QueueInfo,
+        QueueStats,
+    )
+
+JOBS_TABLE = "saq_jobs"
+
+CREATE_JOBS_TABLE = """
+CREATE TABLE IF NOT EXISTS saq_jobs (
+    key TEXT PRIMARY KEY,
```
Comment (tobymao) [inline saq/queue/postgres.py:38]: we'll need some other indices, like something that partitions on status, scheduled, etc

File: saq/queue/postgres.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        max_size: int = 5,
+    ) -> None:
+        super().__init__(name=name, dump=dump, load=load)
+
+        self.pool = pool
+        self.max_size = max_size
+
+    async def connect(self) -> None:
+        await self.pool.open()
+        await self.pool.resize(min_size=0, max_size=self.max_size)
```
Comment (tobymao) [inline saq/queue/postgres.py:83]: what does resize min_size=0 do?

File: saq/queue/postgres.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        if kind == "active":
+            status = [Status.ACTIVE]
+        if kind == "incomplete":
+            status = [Status.NEW, Status.DEFERRED, Status.QUEUED, Status.ACTIVE]
+
+        if status:
+            async with self.pool.connection() as conn, conn.cursor() as cursor:
+                await cursor.execute(
+                    f"""
+                    SELECT count(*) FROM {JOBS_TABLE}
```
Comment (tobymao) [inline saq/queue/postgres.py:127]: should we use %() for everything here?

File: saq/queue/postgres.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+        return []
+
+    async def listen(
+        self,
+        job_keys: Iterable[str],
+        callback: ListenCallback,
+        timeout: float | None = 10,
+    ) -> None:
+        async with self.pool.connection() as conn:
+            await conn.set_autocommit(True)
```
Comment (tobymao) [inline saq/queue/postgres.py:149]: why do you manually neet to set this?

File: saq/queue/postgres.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        logger.info("Finished %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def dequeue(self, timeout: float = 0) -> Job | None:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                f"""
+                UPDATE {JOBS_TABLE} SET status = %(active)s
+                WHERE key = (
+                  SELECT key FROM {JOBS_TABLE}
+                  WHERE status = %(queued)s AND queue = %(queue)s and EXTRACT(EPOCH FROM now()) >= scheduled
```
Comment (tobymao) [inline saq/queue/postgres.py:270]: these are the indices to care about

File: saq/queue/postgres.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                DELETE FROM {JOBS_TABLE}
+                WHERE key = %(key)s
+                """,
+                {"key": key},
+            )
+
+    def _escape_channel_name(self, key: str) -> str:
+        return f"channel_{key.replace('-', '_')}"
+
+    def _unescape_channel_name(self, channel: str) -> str:
```
Comment (tobymao) [inline saq/queue/postgres.py:333]: why do you need this?

File: saq/queue/redis.py [hunk 7]
```diff
@@ -340,13 +340,8 @@ async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
                 await self.redis.lrem(self._active, 0, job.id)
 
     async def retry(self, job: Job, error: str | None) -> None:
+        await super().retry(job, error)
```
Comment (tobymao) [inline saq/queue/redis.py:343]: can we make it so you never have to call super? and then the subclasses implement _retry() or something

File: setup.py [hunk 8]
```diff
@@ -36,11 +36,13 @@
     ],
     extras_require={
         "hiredis": ["redis[hiredis]>=4.2.0"],
+        "postgres": ["psycopg[pool]"],
```
Comment (tobymao) [inline setup.py:39]: let's remove redis from install_requires line 34

File: saq/queue/postgres.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        ListenCallback,
+        DumpType,
+        LoadType,
+        QueueInfo,
+        QueueStats,
+    )
+
+JOBS_TABLE = "saq_jobs"
+
+CREATE_JOBS_TABLE = """
```
Comment (barakalon) [inline saq/queue/postgres.py:36]: 1. Some applications might not have permission to create tables, so this should probably be made optional and executable by a standalone process
2. Should we be thinking ahead about migrations? What if we want to add a column, etc. Perhaps we should be syncing a sequence of changes.

Comment (barakalon) [reply saq/queue/postgres.py:36]: Backwards compatible migrations, that don’t require users to go in and alter their DB, that is

Comment (tobymao) [reply saq/queue/postgres.py:36]: i think it's ok that we always create the table, but i'd make it a tad more configurable like what the table name is?

Comment (barakalon) [reply saq/queue/postgres.py:36]: So if we ever need to change we expect users to update the table name?

File: saq/queue/postgres.py (hunk 6 above)
Comment (vchan) [reply saq/queue/postgres.py:333]: Postgres channel names have the same restrictions as table names. This was a hack to work around it. I'll add a better way to encode channel names.

File: saq/queue/postgres.py (hunk 2 above)
Comment (vchan) [reply saq/queue/postgres.py]: For AsyncNullConnectionPool, min_size is always 0 since it doesn't maintain any connections. I've switched the queue to use an AsyncConnectionPool.

File: saq/queue/postgres.py (hunk 6 above)
Comment (vchan) [reply saq/queue/postgres.py:343]: Realized I could just quote the channel name and not worry about this...

File: saq/queue/postgres.py (hunk 4 above)
Comment (vchan) [reply saq/queue/postgres.py:164]: Due to the way notifications are implemented in Postgres, if I don't set this, the LISTEN will be in a transaction and won't receive any notifications until the transaction ends.

File: saq/queue/base.py [hunk 10]
```diff
@@ -147,37 +143,91 @@ async def enqueue(self, job_or_func: str | Job, **kwargs: t.Any) -> Job | None:
...
-
-        job_dict = self._load(job_bytes)
-        if job_dict.pop("queue") != self.name:
-            raise ValueError(f"Job {job_dict} fetched by wrong queue: {self.name}")
-        return Job(**job_dict, queue=self)
+    async def sweep(self, lock: int = 60, abort: float = 5.0) -> list[t.Any]:
+        job_ids = await self._get_active_jobs_for_sweep(lock)
+
+        swept = []
+        if job_ids:
```
Comment (tobymao) [inline saq/queue/base.py:193]: this is redis specific hack, i don't think we need this for postgres right

File: saq/queue/postgres.py [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+    def deserialize(self, job_dict: dict[t.Any, t.Any]) -> Job | None:
+        if job_dict.pop("queue") != self.name:
+            raise ValueError(f"Job {job_dict} fetched by wrong queue: {self.name}")
+        return Job(**job_dict, queue=self)
+
+    async def disconnect(self) -> None:
+        await self.pool.close()
+
+    async def info(
+        self, jobs: bool = False, offset: int = 0, limit: int = 10
```
Comment (tobymao) [inline saq/queue/postgres.py:117]: doesn't look like jobs, offset, or limit is used

File: saq/queue/postgres.py (hunk 3 above)
Comment (tobymao) [reply saq/queue/postgres.py:152]: shouldn't this be self.jobs_table?

File: saq/queue/postgres.py [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+                    SELECT count(*) FROM {JOBS_TABLE}
+                    WHERE status = 'active'
+                      AND queue = %(queue)s
+                    """,
+                    {"queue": self.name},
+                )
+            elif kind == "incomplete":
+                await cursor.execute(
+                    f"""
+                    SELECT count(*) FROM {JOBS_TABLE}
```
Comment (tobymao) [inline saq/queue/postgres.py:161]: ditto

File: saq/queue/postgres.py [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+            if job:
+                return self.deserialize(job[0])
+        return None
+
+    async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
+        job.error = error
+        job.status = Status.ABORTING
+
+        await self.update(job)
+        asyncio.create_task(self.delete(job, status=Status.ABORTED, delay=ttl))
```
Comment (tobymao) [inline saq/queue/postgres.py:239]: why create_task here?

where is the code that actually kills the task?

File: saq/queue/postgres.py [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+                    f"""
+                    UPDATE {self.jobs_table} SET status = 'active'
+                    WHERE key = (
+                      SELECT key
+                      FROM {self.jobs_table}
+                      WHERE status = 'queued'
+                        AND queue = %(queue)s
+                        AND %(now)s >= (queued + scheduled)
+                      FOR UPDATE SKIP LOCKED
+                      LIMIT 1
```
Comment (tobymao) [inline saq/queue/postgres.py:326]: shouldn't limit 1 be determined by the concurrency argument?

File: saq/queue/postgres.py [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+            if not await cursor.fetchone():
+                return None
+
+        logger.info("Enqueuing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+        return job
+
+    async def delete(
+        self, job: Job, status: Status | None = None, delay: float = 0
+    ) -> None:
+        if delay:
```
Comment (tobymao) [inline saq/queue/postgres.py:383]: why do we need a delay for delete?

File: saq/queue/postgres.py (hunk 12 above)
Comment (vchan) [reply saq/queue/postgres.py:161]: That's a typo.

File: saq/queue/postgres_ddl.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+    queued BIGINT NOT NULL DEFAULT EXTRACT(EPOCH FROM now()),
+    job JSONB,
+    queue TEXT,
+    status TEXT,
+    scheduled BIGINT
+);
+"""
+
+CREATE_JOBS_DEQUEUE_INDEX = """
+CREATE INDEX IF NOT EXISTS saq_jobs_dequeue_idx ON {jobs_table} (status, queue, (queued + scheduled))
```
Comment (tobymao) [inline saq/queue/postgres_ddl.py:14]: does this index work? queued + scheduled? does that get hit?

File: saq/queue/postgres.py (hunk 13 above)
Comment (vchan) [reply saq/queue/postgres.py:239]: This deletes aborted jobs from the jobs table. The worker contains the code that kills the task.

File: saq/queue/postgres_ddl.py (hunk 16 above)
Comment (tobymao) [reply saq/queue/postgres_ddl.py:14]: why do we have to do math? is that how it was done? is scheduled not just a timestamp it's an integer seconds?

File: saq/queue/postgres.py (hunk 13 above)
Comment (vchan) [reply saq/queue/postgres.py:239]: I think we need a janitor task to clear out jobs from the jobs table based on ttl, so this can go away.

File: saq/queue/postgres.py (hunk 15 above)
Comment (vchan) [reply saq/queue/postgres.py:383]: This is tied to the ttl for removing completed (including aborted) tasks. I'll add a janitor task to replace this though.

File: saq/queue/postgres.py [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+
+
+class PostgresQueue(Queue):
+    """
+    Queue is used to interact with Postgres.
+
+    Args:
+        pool: instance of psycopg_pool.AsyncConnectionPool
+        name: name of the queue (default "default")
+        joabs_table: name of the Postgres table SAQ will write jobs to (default "saq_jobs")
```
Comment (tobymao) [inline saq/queue/postgres.py:54]: jobs_table

File: saq/queue/postgres.py [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+            json.set_json_loads(load)
+
+        self.connections: dict[str, AsyncConnection] = {}
+        self.cond = asyncio.Condition()
+
+    async def connect(self) -> None:
+        await self.pool.open()
+        await self.pool.resize(min_size=self.min_size, max_size=self.max_size)
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(CREATE_JOBS_TABLE.format(jobs_table=self.jobs_table))
```
Comment (tobymao) [inline saq/queue/postgres.py:103]: can you do a for loop on all of the init scripts?

File: saq/queue/postgres.py [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+                f"""
+                SELECT job
+                FROM {self.jobs_table} LEFT OUTER JOIN pg_locks ON lock_id = objid
+                WHERE status = 'active'
+                  AND objid IS NULL
+                """
+            )
+            results = await cursor.fetchall()
+        swept = []
+        for result in results:
```
Comment (tobymao) [inline saq/queue/postgres.py:218]: what happens if the sweeper gets killed after the delete from jobs table but before retry / abort?

File: saq/queue/postgres.py [hunk 20]
```diff
@@ -0,0 +1,521 @@
...
+    async def _notify(self, channel: str, payload: t.Any) -> None:
+        async with self.pool.connection() as conn:
+            await conn.execute(f"NOTIFY \"{channel}\", '{payload}'")
+
+    async def update(self, job: Job) -> None:
+        job.touched = now()
+        async with self._get_connection(job.key) as conn:
+            await conn.execute(
+                f"""
+                    UPDATE {self.jobs_table} SET job = %(job)s, status = %(status)s
```
Comment (tobymao) [inline saq/queue/postgres.py:268]: are you unable to pass in self.jobs_table as a param? what if thehre's sql injection in there

File: saq/queue/postgres.py [hunk 21]
```diff
@@ -0,0 +1,521 @@
...
+            )
+            job = await cursor.fetchone()
+            if job:
+                return self.deserialize(job[0])
+        return None
+
+    async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
+        job.error = error
+        job.status = Status.ABORTING
+        job.ttl = int(seconds(now()) + ttl)
```
Comment (tobymao) [inline saq/queue/postgres.py:292]: why are you setting the job ttl here? i don't think redis does this. redis usses the abort ttl for the abort key right?

File: saq/queue/postgres.py [hunk 22]
```diff
@@ -0,0 +1,521 @@
...
+        self._update_job_for_retry(job, error)
+        next_retry_delay = job.next_retry_delay()
+        if next_retry_delay:
+            scheduled = time.time() + next_retry_delay
+        else:
+            scheduled = job.scheduled or seconds(now())
+
+        async with self._get_connection(job.key) as conn:
+            await conn.execute(
+                f"""
```
Comment (tobymao) [inline saq/queue/postgres.py:306]: should this just leverage the update code? and update should better handle passing in scheduled?

Comment (tobymao) [reply saq/queue/postgres.py:306]:
```
def update(self, job, status=..., scheduled=...)
```

File: saq/queue/postgres.py [hunk 23]
```diff
@@ -0,0 +1,521 @@
...
+        await job.finish(Status.ABORTED, error=job.error)
+
+    async def dequeue_timer(self, poll_interval: int) -> None:
+        """Wakes up a single dequeue task every `poll_interval` seconds."""
+        while True:
+            async with self.cond:
+                self.cond.notify(1)
+            await asyncio.sleep(poll_interval)
+
+    async def wait_for_job(self) -> Job:
```
Comment (tobymao) [inline saq/queue/postgres.py:432]: i feel like this isn't necessary, it's pretty much the same as dequeue, can we move it inline into dequeu?

File: saq/queue/postgres.py [hunk 24]
```diff
@@ -0,0 +1,521 @@
...
+                await self.cond.wait()
+            job = await self._dequeue()
+            if job:
+                break
+        return job
+
+    async def listen_for_enqueues(self, timeout: float | None = None) -> None:
+        """Wakes up a single dequeue task when a Postgres enqueue notification is received."""
+        async with self.pool.connection() as conn:
+            await conn.execute('LISTEN "saq:enqueue"')
```
Comment (tobymao) [inline saq/queue/postgres.py:449]: saq:enqueue make constant?

File: saq/queue/postgres.py (hunk 23 above)
Comment (vchan) [reply saq/queue/postgres.py:454]: Do you mean make this an inner function in dequeue()? This was added so dequeue() can use asyncio.wait_for() for timing out, but this can be nested inside dequeue().

Comment (tobymao) [reply saq/queue/postgres.py:454]: yea lambda inside

File: saq/queue/postgres.py (hunk 21 above)
Comment (vchan) [reply saq/queue/postgres.py:322]: Redis also sets the job expiration to abort ttl + 1. I can update this to match what Redis is doing.

File: saq/queue/postgres.py (hunk 19 above)
Comment (vchan) [reply saq/queue/postgres.py]: The delete above is for clearing out jobs with terminal statuses past their ttl, so it's not related to retry/abort. However, all of sweep now takes place in a transaction, so everything should be rolled back if the sweeper gets killed before it finishes.

File: saq/queue/postgres.py [hunk 25]
```diff
@@ -0,0 +1,608 @@
...
+    async def sweep(self, lock: int = 60, abort: float = 5.0) -> list[t.Any]:
+        """Delete jobs past their TTL from the jobs table"""
+        swept = []
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                """
+            SELECT pg_try_advisory_xact_lock(%(classid)s, %(objid)s)
+            """,
+                {"classid": SWEEP_LOCK_KEY[0], "objid": SWEEP_LOCK_KEY[1]},
```
Comment (tobymao) [inline saq/queue/postgres.py:271]: why is the sweep lock key 1, 1, how did you decide that?

File: saq/queue/postgres.py [hunk 26]
```diff
@@ -0,0 +1,608 @@
...
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"now": math.ceil(seconds(now()))},
+            )
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT job
+                FROM {jobs_table} LEFT OUTER JOIN pg_locks ON lock_key = objid
+                WHERE status = 'active'
```
Comment (tobymao) [inline saq/queue/postgres.py:293]: so this finds jobs that are active but not locked, so you know they are stuck right?

File: saq/queue/postgres.py [hunk 27]
```diff
@@ -0,0 +1,608 @@
...
+        self, job: Job, status: Status | None = None, scheduled: int | None = None
+    ) -> None:
+        if status:
+            job.status = status
+        if scheduled:
+            job.scheduled = scheduled
+        job.touched = now()
+
+        async with self.pool.connection() as conn:
+            await conn.execute(
```
Comment (tobymao) [inline saq/queue/postgres.py:355]: is this done in a transaction? can we also notify in the transaction too so that you'll never update a job and skip a notification?

File: saq/queue/postgres.py [hunk 28]
```diff
@@ -0,0 +1,608 @@
...
+                {"key": job_key},
+            )
+            job = await cursor.fetchone()
+            if job:
+                return self.deserialize(job[0])
+        return None
+
+    async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
+        job.error = error
+        job.status = Status.ABORTING
```
Comment (tobymao) [inline saq/queue/postgres.py:389]: where is the job actually aborted?

File: saq/queue/postgres.py [hunk 29]
```diff
@@ -0,0 +1,608 @@
...
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                INSERT INTO {jobs_table} (key, job, queue, status, scheduled)
+                VALUES (%(key)s, %(job)s, %(queue)s, %(status)s, %(scheduled)s)
+                ON CONFLICT (key) DO UPDATE
+                SET job = %(job)s, queue = %(queue)s, status = %(status)s, scheduled = %(scheduled)s
+                WHERE {jobs_table}.status NOT IN ('queued', 'aborting')
```
Comment (tobymao) [inline saq/queue/postgres.py:490]: so we won't update jobs that are queued or aborting, but we will update them if they are active???

File: saq/queue/postgres.py [hunk 30]
```diff
@@ -0,0 +1,608 @@
...
+        return job
+
+    async def get_abort_errors(self, jobs: Iterable[Job]) -> list[bytes | None]:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT key, job->'error'
+                FROM {jobs_table}
+                WHERE key = ANY(%(keys)s)
```
Comment (tobymao) [inline saq/queue/postgres.py:518]: is this always sorted automatically? does the order matter?

File: saq/queue/postgres.py [hunk 31]
```diff
@@ -0,0 +1,608 @@
...
+                  ORDER BY scheduled
+                  LIMIT %(limit)s
+                  FOR UPDATE SKIP LOCKED
+                )
+                UPDATE {jobs_table} SET status = 'active'
+                FROM locked_job
+                WHERE {jobs_table}.key = locked_job.key AND pg_try_advisory_lock(locked_job.lock_key)
+                RETURNING job
+                """
+                ).format(jobs_table=self.jobs_table),
```
Comment (tobymao) [inline saq/queue/postgres.py:573]: why is jobs table passed in with format and not the params below? we shouldn't use old style over f-strings, that's not what i meant

File: saq/queue/postgres.py [hunk 32]
```diff
@@ -0,0 +1,608 @@
...
+            return
+        async with self._get_connection() as conn:
+            await conn.execute(
+                SQL(
+                    """
+                SELECT pg_advisory_unlock(lock_key)
+                FROM {jobs_table}
+                WHERE key = ANY(%(keys)s)
+                """
+                ).format(jobs_table=self.jobs_table),
```
Comment (tobymao) [inline saq/queue/postgres.py:604]: same here

File: saq/queue/postgres.py [hunk 33]
```diff
@@ -0,0 +1,608 @@
...
+                RETURNING key
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"now": math.ceil(seconds(now()))},
+            )
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT job
+                FROM {jobs_table} LEFT OUTER JOIN pg_locks ON lock_key = objid
```
Comment (tobymao) [inline saq/queue/postgres.py:292]: how does lock_key = objid if objid is null? null never == null

File: saq/queue/postgres.py [hunk 34]
```diff
@@ -0,0 +1,608 @@
...
+    async def stats(self, ttl: int = 60) -> QueueStats:
+        stats = self._get_stats()
+        async with self.pool.connection() as conn:
+            await conn.execute(
+                SQL(
+                    """
+                INSERT INTO {stats_table} (worker_id, stats, ttl)
+                VALUES (%(worker_id)s, %(stats)s, %(ttl)s)
+                ON CONFLICT (worker_id) DO UPDATE
+                SET stats = %(stats)s, ttl = %(ttl)s
```
Comment (tobymao) [inline saq/queue/postgres.py:206]: is the stats table ever cleaned up? what does the ttl do?

Comment (vchan) [reply saq/queue/postgres.py:206]: TTL is used by info to filter out stale stats. I'll also add something in the sweeper to delete old stats.

File: saq/queue/postgres.py (hunk 25 above)
Comment (vchan) [reply saq/queue/postgres.py:271]: I just picked it arbitrarily. I didn't want it to potentially collide with the job lock keys.

File: saq/queue/postgres.py (hunk 33 above)
Comment (vchan) [reply saq/queue/postgres.py:292]: Right, but there could be null objids because it's a left outer join if some jobs don't have locks.

File: saq/queue/postgres.py (hunk 26 above)
Comment (vchan) [reply saq/queue/postgres.py:293]: Yes, exactly.

File: saq/queue/postgres.py (hunk 34 above)
Comment (tobymao) [reply saq/queue/postgres.py:206]: what does redis do?

File: saq/queue/postgres.py (hunk 28 above)
Comment (vchan) [reply saq/queue/postgres.py:389]: The worker's abort function cancels the task and calls queue.finish_abort(), which updates the status to aborted.

File: saq/queue/postgres.py (hunk 27 above)
Comment (tobymao) [reply saq/queue/postgres.py:355]: can you look through all the places and make sure things are as robust as possible with using transactions?

File: saq/queue/postgres.py (hunk 31 above)
Comment (vchan) [reply saq/queue/postgres.py:573]: This is calling format on a psycopg.sql.SQL object, not on a string. This is the psycopg way to dynamically set tables and fields for a query and prevent SQL injection. We can't bind tables and fields in queries like other query values with placeholders like with the params below.

Comment (tobymao) [reply saq/queue/postgres.py:573]: great

File: saq/queue/postgres.py (hunk 29 above)
Comment (vchan) [reply saq/queue/postgres.py:490]: Good point. I think we should only update, aka re-queue, jobs that are done, i.e. finished, failed, or aborted. What do you think?

Comment (tobymao) [reply saq/queue/postgres.py:490]: TERMINAL_STATUSES right?

File: saq/queue/postgres.py (hunk 34 above)
Comment (vchan) [reply saq/queue/postgres.py:206]: The Redis queue:

1. Saves the current worker's stat to key "saq:stats:{worker_id}" with an expiration of ttl.
2. Removes stale worker_ids in the sorted set "saq:stats".
3. Add the current worker's id to the sorted set "saq:stats".
4. Sets an expiration of ttl on the sorted set "saq:stats".

Since we can't automatically expire rows in our stats tables, I'm using ttl to filter out stale stats in info() and add a step in the sweeper to delete them.

File: saq/queue/postgres.py (hunk 29 above)
Comment (vchan) [reply saq/queue/postgres.py:490]: Yes, TERMINAL_STATUSES

File: saq/queue/postgres.py (hunk 30 above)
Comment (vchan) [reply saq/queue/postgres.py:518]: Yes, order does matter. No, I don't think order is guaranteed here. There's additional logic below to reorder the errors to match the input jobs list.

File: saq/queue/postgres.py [hunk 35]
```diff
@@ -0,0 +1,648 @@
...
+
+        self.cond = asyncio.Condition()
+        self.queue: asyncio.Queue = asyncio.Queue()
+        self.waiting = 0  # Internal counter of worker tasks waiting for dequeue
+        self._connected = False
+        self.connection_lock = asyncio.Lock()
+        self.released: list[str] = []
+
+    async def connect(self) -> None:
+        if self._connected:
```
Comment (tobymao) [inline saq/queue/postgres.py:111]: let's remove self._connected and look at self.connection or self.pool instead

File: saq/queue/postgres.py (hunk 25 above)
Comment (tobymao) [reply saq/queue/postgres.py:271]: it doesn't seem like you're using the lock timer here, we only want to sweep every 60 seconds no matter how many workers there are right

File: saq/queue/postgres.py (hunk 26 above)
Comment (tobymao) [reply saq/queue/postgres.py:311]: what about jobs that are stuck in 'aborting' ?

File: saq/queue/postgres.py [hunk 36]
```diff
@@ -0,0 +1,648 @@
...
+        status: Status,
+        *,
+        result: t.Any = None,
+        error: str | None = None,
+    ) -> None:
+        self._update_job_for_finish(job, status, result=result, error=error)
+        key = job.key
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            if job.ttl >= 0:
```
Comment (tobymao) [inline saq/queue/postgres.py:457]: hey, this doesn't match redis's behavior, if ttl > 0, then set expiration, if it's 0, there is no expiration, else delete. right?

File: saq/queue/redis.py [hunk 37]
```diff
@@ -392,12 +392,7 @@ async def finish(
...
 
-            if status == Status.COMPLETE:
-                self.complete += 1
-            elif status == Status.FAILED:
-                self.failed += 1
-            elif status == Status.ABORTED:
-                self.aborted += 1
+            self._update_stats(status)
 
             await self.notify(job)
```
Comment (tobymao) [inline saq/queue/redis.py:397]: why is conn passed in some places but not others?

Comment (vchan) [reply saq/queue/redis.py:397]: This is the Redis queue, so there's no conn to pass here. In Postgres, I've centralized most calls to notify() in update() and the remaining call to notify() in finish() passes conn. We generally want conn so notify can be part of conn's transaction.

File: saq/queue/postgres.py (hunk 36 above)
Comment (vchan) [reply saq/queue/postgres.py:456]: Good catch. This is wrong after I refactored finish() to use update(). I've fixed it.

File: saq/queue/postgres.py [hunk 38]
```diff
@@ -439,7 +439,6 @@ async def retry(self, job: Job, error: str | None) -> None:
         await self._release_job(job.key)
 
         self.retried += 1
-        await self.notify(job)
```
Comment (tobymao) [inline saq/queue/postgres.py]: why is this removed?

Comment (vchan) [reply saq/queue/postgres.py]: The update() function already calls notify().

File: saq/queue/postgres.py (hunk 25 above)
Comment (vchan) [reply saq/queue/postgres.py:271]: Ah, you're right. Need to update this.

File: saq/queue/postgres.py [hunk 39]
```diff
@@ -0,0 +1,648 @@
...
+                    """
+                DELETE FROM {jobs_table}
+                WHERE status IN ('aborted', 'complete', 'failed')
+                  AND %(now)s >= ttl
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"now": math.ceil(seconds(now()))},
+            )
+            # Delete expired stats
+            await cursor.execute(
```
[the remark this answers is not in the document]
Comment (vchan) [reply saq/queue/postgres.py]: Yes, I've update this to combine them now.

File: saq/queue/postgres.py (hunk 26 above)
Comment (vchan) [reply saq/queue/postgres.py]: Yes, if the worker aborting the job dies before it can update the job's status. I'll change this to also sweep jobs stuck in 'aborting`.

Comment (tobymao) [reply saq/queue/postgres.py]: make sure we don’t sweep jobs that are actually aborting

Comment (vchan) [reply saq/queue/postgres.py]: Okay, I've double checked the sweeper makes sure there's no lock associated with a job before sweeping, and aborting releases the lock after canceling the task and updating the status to 'aborted'.

File: setup.py [hunk 40]
```diff
@@ -31,17 +31,20 @@
...
+        "redis": ["redis>=4.2,<6.0"],
         "web": ["aiohttp", "aiohttp_basicauth"],
         "dev": [
             "black",
             "coverage",
             "mypy",
+            "psycopg[pool]",
```
Comment (barakalon) [inline setup.py:45]: We should just update the `make deps` command to include redis and postgres: https://github.com/tobymao/saq/blob/master/Makefile#L2

File: tests/helpers.py [hunk 41]
```diff
@@ -1,13 +1,32 @@
...
+    await queue.connect()
+    return queue
+
+
+async def cleanup_queue(queue: Queue) -> None:
+    if isinstance(queue, RedisQueue):
+        await queue.redis.flushdb()
+    elif isinstance(queue, PostgresQueue):
+        async with queue.pool.connection() as conn:
+            await conn.execute("DROP SCHEMA public CASCADE")
```
Comment (barakalon) [inline tests/helpers.py:28]: Yikes. Should we be using a dedicated test schema?

File: saq/queue/postgres.py [hunk 42]
```diff
@@ -0,0 +1,667 @@
...
+        self.has_sweep_lock = False
+
+    async def info(
+        self, jobs: bool = False, offset: int = 0, limit: int = 10
+    ) -> QueueInfo:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT worker_id, stats FROM {stats_table}
```
[the remark this answers is not in the document]
Comment (vchan) [reply saq/queue/postgres.py]: I passed all of the multiline SQL strings to `textwrap.dedent`. Wanted to confirm that is what you meant.
